### ISSUE TITLE
[BUG] Call to a get_or_init inside a get_or_init cause a stack overflow

### ISSUE THREAD

**Issue by TxMat** [danieldg/async-once-cell#13]
When getting the content of an AsyncOnceCell inside the init method of another the program panics with the error 

```sh
thread 'tests::test_get_user_id_from_email' has overflowed its stack
fatal runtime error: stack overflow
```

Erroneous code :

```rust
mod tests {
    use async_once_cell::OnceCell;
    use aws_config::BehaviorVersion;
    use aws_config::meta::region::RegionProviderChain;
    use aws_sdk_dynamodb::Client as DynamoDbClient;
    use aws_sdk_dynamodb::types::AttributeValue;
    
    static DYNAMODB_CLIENT: OnceCell<DynamoDbClient> = OnceCell::new();
    static INIT_TEST_USER: OnceCell<bool> = OnceCell::new();
    
    async fn init_dynamodb_client() -> DynamoDbClient {
        println!("initializing dynamodb client");
        let region_provider = RegionProviderChain::default_provider().or_else("eu-west-2");
        let config = aws_config::defaults(BehaviorVersion::v2023_11_09())
            .region(region_provider)
            .load()
            .await;
        aws_sdk_dynamodb::Client::new(&config)
    }
    
    async fn get_dynamodb_client() -> DynamoDbClient {
        println!("getting dynamodb client");
        DYNAMODB_CLIENT.get_or_init(async {
            init_dynamodb_client().await
        }).await.clone()
    }
    
    async fn insert_test_user_to_db() {
        println!("inserting test user");
        INIT_TEST_USER.get_or_init(async {
            let dynamodb_client = get_dynamodb_client().await;
            //// Error happens here ////
            // snip
            true
        }).await;
    }
    
    #[tokio::test]
    async fn test_get_user_id_from_email() {
        println!("testing get user id from email");
        insert_test_user_to_db().await;
        // snip
    }
}
```

**Comment by danieldg** [comment]
This may be a rustc bug (excessive stack use in an async function); it's not specific to async_once_cell.  At least some rustc versions don't always overlap arguments with their uses, which might end up magnifying the size of some of your futures.  A test with `std::mem::size_of_val` called on the result of some of these functions might help you narrow down what is actually happening: ideally the size of `let f = insert_test_user_to_db();` should be a few hundred bytes larger than `init_dynamodb_client`, but it's possible that it is instead a multiple of the size.

It might be possible to mitigate this by rewriting some of async_once_cell's functions, but I'd prefer to see a short example that shows the problem (use something like  `[u8;1000]` held across an await to show it's multiplying sizes) - ideally one that can be put in a unit test.

**Comment by TxMat** [comment]
The error might be with the dynamodb client somehow, 

```rust
lazy_static!(
        static ref DYNAMODB_CLIENT: OnceCell<DynamoDbClient> = OnceCell::new();
        static ref INIT_TEST_USER: OnceCell<bool> = OnceCell::new();
    );


    async fn init_dynamodb_client() -> DynamoDbClient {
        println!("initializing dynamodb client");
        let region_provider = RegionProviderChain::default_provider().or_else("eu-west-2");
        let config = aws_config::defaults(BehaviorVersion::v2023_11_09())
            .region(region_provider)
            .load()
            .await;
        aws_sdk_dynamodb::Client::new(&config)
    }

    async fn get_dynamodb_client() -> DynamoDbClient {
        println!("getting dynamodb client");
        DYNAMODB_CLIENT.get_or_init(async {
            init_dynamodb_client().await
        }).await.clone()
    }

    async fn insert_test_user_to_db() {
        println!("inserting test user");
        INIT_TEST_USER.get_or_init(async {
            let dynamodb_client = get_dynamodb_client().await;
            //// Error happens here ////
            // snip
            true
        }).await;
    }

    #[tokio::test]
    async fn test_get_user_id_from_email() {
        println!("testing get user id from email");
        let f = insert_test_user_to_db();
        println!("f: {:?}", std::mem::size_of_val(&f));
        // f.await;
        // snip
    }
```

```shell
f: 339704
```

the size does not seem to magnify with an [u8;1000]

```rust
lazy_static!(
        static ref DYNAMODB_CLIENT: OnceCell<[u8;1000]> = OnceCell::new();
        static ref INIT_TEST_USER: OnceCell<bool> = OnceCell::new();
    );


    async fn init_dummy_data() -> [u8;1000] {
        println!("initializing dummy data");
        [0;1000]
    }

    async fn get_dummy_data() -> [u8;1000] {
        println!("getting data");
        DYNAMODB_CLIENT.get_or_init(async {
            init_dummy_data().await
        }).await.clone()
    }

    async fn do_someting_with_data() {
        println!("inserting test user");
        INIT_TEST_USER.get_or_init(async {
            let data = get_dummy_data().await;
            //// Error happens here ////
            // snip
            true
        }).await;
    }

    #[tokio::test]
    async fn test_get_user_id_from_email() {
        println!("testing get user id from email");
        let f = do_someting_with_data();
        println!("f: {:?}", std::mem::size_of_val(&f));
        f.await;
        // snip
    }
```

```shell
f: 904
```

Unfortunately, i'm still pretty new to rust and unable to extrapolate the results but i guess it's the AWS SDK's fault

**Comment by Mossop** [comment]
I'm afraid I can't provide a short example but just throwing in my experience. I've been using `OnceCell` to cache the result of operations that may be needed by parallel tasks. For a while I've been plagued with stack overflows and being unable to diagnose have been slowly increasing the per thread stack size. Once I hit 30MB and it was still overflowing and I had something semi reproducible I narrowed it down to a call to `OnceCell::get_or_try_init` which led me to finding this issue. I ended up switching to tokio's `OnceCell` and with no other changes the problem went away and I was able to drop the stack size back down to 2MB.

**Comment by danieldg** [comment]
If the error is due to rustc not unifying futures, then e0565dababbbff57f86f7c7b2620f5d718c66b47 should address this. @Mossop or @TxMat, are you able to test that?

**Comment by Mossop** [comment]
It's definitely an improvement. I only had to increase the stack limit to around 7MB for it to work in my case. For some reason with the current 0.5.3 release without even creating a `async_once_cell::OnceCell` my app won't run with a 2MB stack limit right now.

I did attempt to pull out what I thought were the relevant parts of my code into a simplified example but it didn't fail.

**Comment by danieldg** [comment]
I expect that this change (which is now published as 0.5.4) is about as good as it can get without either rustc itself improving field overlapping or pulling out all uses of `async fn` in favor of pin-projection structs.  Workarounds like calling `Box::pin(async { ... }).await` or pinning a future in the caller's stack frame and passing `Pin<&mut F>` as the init argument can reduce the Future object size further at the expense of cluttering call-sites.
